lib/_http_parser.js

```diff
@@ -0,0 +1,879 @@
+    some kind of "big integer" library for lengths > 2^53-1.
+
+  * No special handling for `Proxy-Connection` header. The reason for this is
+    that `Proxy-Connection` was an experimental header for HTTP 1.0 user agents
+    that ended up being a bad idea because of the confusion it can bring. You
+    can read a bit more about this in
+    [RFC 7230 A.1.2](https://tools.ietf.org/html/rfc7230#page-79)
+*/
+
+var inspect = require('util').inspect;
+
+var CR = 13;
+var LF = 10;
+
+var MAX_CHUNK_SIZE = Number.MAX_SAFE_INTEGER; // 9007199254740991
+
+// RFC 7230 recommends HTTP implementations support at least 8000 bytes for
+// the request line. We use 8190 by default, the same as Apache.
+HTTPParser.MAX_REQ_LINE = 8190;
+// RFC 7230 does not have any recommendations for minimum response line length
+// support. Judging by the (current) longest standard status reason text, the
+// typical response line will be 44 bytes or less (not including (CR)LF). Since
+// the reason text field is free form though, we will roughly triple that
+// amount for the default.
+HTTPParser.MAX_RES_LINE = 128;
+
+// This is the total limit for start line + all headers was copied from
+// joyent/http-parser.
+var MAX_HEADER_BYTES = 80 * 1024;
+
+var RE_CONN_CLOSE = /(?:^|[\t ,]+)close(?:\r?$||[\t ,]+)/i;
+var RE_CONN_KEEPALIVE = /(?:^|[\t ,]+)keep\-alive(?:\r?$|[\t ,]+)/i;
+var RE_CONN_UPGRADE = /(?:^|[\t ,]+)upgrade(?:\r?$|[\t ,]+)/i;
+var RE_TE_CHUNKED = /(?:^|[\t ,]+)chunked(?:\r?$|[\t ,]+)/i;
+var CC_CONNECT = 'connect'.split('').map(getFirstCharCode);
+var CC_CONNECTION = 'connection'.split('').map(getFirstCharCode);
+var CC_TE = 'transfer-encoding'.split('').map(getFirstCharCode);
+var CC_UPGRADE = 'upgrade'.split('').map(getFirstCharCode);
+var CC_CONTLEN = 'content-length'.split('').map(getFirstCharCode);
+var REQ_HTTP_VER_BYTES = ' HTTP/1.'.split('').map(getFirstCharCode);
+var RES_HTTP_VER_BYTES = REQ_HTTP_VER_BYTES.slice(1);
+REQ_HTTP_VER_BYTES.reverse();
+
+var STATE_REQ_LINE = 0;
+var STATE_STATUS_LINE = 1;
+var STATE_HEADER = 2;
+var STATE_BODY_LITERAL = 3;
+var STATE_BODY_EOF = 4;
+var STATE_BODY_CHUNKED_SIZE = 5;
+var STATE_BODY_CHUNKED_SIZE_IGNORE = 6;
+var STATE_BODY_CHUNKED_BYTES = 7;
+var STATE_BODY_CHUNKED_BYTES_LF = 8;
+var STATE_COMPLETE = 9;
+
+var FLAG_CHUNKED = 1 << 0;
+var FLAG_CONNECTION_KEEP_ALIVE = 1 << 1;
+var FLAG_CONNECTION_CLOSE = 1 << 2;
+var FLAG_CONNECTION_UPGRADE = 1 << 3;
+var FLAG_TRAILING = 1 << 4;
+var FLAG_UPGRADE = 1 << 5;
+var FLAG_SKIPBODY = 1 << 6;
+var FLAG_SHOULD_KEEP_ALIVE = 1 << 7;
+var FLAG_CONNECT_METHOD = 1 << 8;
+var FLAG_ANY_UPGRADE = FLAG_UPGRADE | FLAG_CONNECTION_UPGRADE;
+
+function HTTPParser(type) {
+  this.onHeaders = null;
+  this.onBody = null;
+  this.onComplete = null;
+
+  // Extra stuff tagged onto parser object by core modules/files
+  this.onIncoming = null;
+  this.incoming = null;
+  this.socket = null;
+
+  this.reinitialize(type);
```

Comment:
The reason for putting initialization in the constructor is so that V8 will put the properties on the object itself, instead of the out-of-band storage.